pkg/servicelb/controller.go

```diff
@@ -260,13 +263,67 @@ func (h *handler) podIPs(pods []*core.Pod) ([]string, error) {
 		ips = keys(intIPs)
 	}
 
+	ips, err := filterByIPFamily(ips, svc)
+	if err != nil {
+		return nil, err
+	}
+
 	if len(ips) > 0 && h.rootless {
 		return []string{"127.0.0.1"}, nil
 	}
 
 	return ips, nil
 }
 
+// filterByIPFamily filters ips based on dual-stack parameters of th service
+func filterByIPFamily(ips []string, svc *core.Service) ([]string, error) {
+
+	var ipv4Addresses []string
+	var ipv6Addresses []string
+
+	for _, ip := range ips {
+		if utilsnet.IsIPv4String(ip) {
+			ipv4Addresses = append(ipv4Addresses, ip)
+		}
+		if utilsnet.IsIPv6String(ip) {
+			ipv6Addresses = append(ipv6Addresses, ip)
+		}
+	}
+
+	switch *svc.Spec.IPFamilyPolicy {
+	case core.IPFamilyPolicySingleStack:
+		if svc.Spec.IPFamilies[0] == core.IPv4Protocol {
+			return ipv4Addresses, nil
+		}
+		if svc.Spec.IPFamilies[0] == core.IPv6Protocol {
+			return ipv6Addresses, nil
+		}
+	case core.IPFamilyPolicyPreferDualStack:
+		if svc.Spec.IPFamilies[0] == core.IPv4Protocol {
+			ipAddresses := append(ipv4Addresses, ipv6Addresses...)
+			return ipAddresses, nil
+		}
+		if svc.Spec.IPFamilies[0] == core.IPv6Protocol {
+			ipAddresses := append(ipv6Addresses, ipv4Addresses...)
+			return ipAddresses, nil
+		}
+	case core.IPFamilyPolicyRequireDualStack:
+		if (len(ipv4Addresses) == 0) || (len(ipv6Addresses) == 0) {
+			return nil, errors.New("IPFamilyPolicy=RequireDualStack but no ipv6 IP defined for the service")
```

Comment:
Errors should not be capitalized; also this could be triggered (unlikely) if we weren't able to find any IPv4 addresses.
```suggestion
			return nil, errors.New("one or more IP families did not have addresses available for service with ipFamilyPolicy=RequireDualStack")
```

Reply:
ok, thanks!